### DISCUSSION TITLE
DAX Studio Export to CSV for datetime columns arrive in XLS as general

### DISCUSSION THREAD

**Discussion by JHCairn** [DaxStudio/DaxStudio#829] [Q&A] [answered]
I am running some queries in DAX Studio to Grid to check them.
Then I output them to static CSV using File Output.
They appear in the grid as date time. Hovering on the fields in DAX Studio also indicates they are DateTime fields, but in the CSV file they exported to, these columns come as 'General' format and the users have to select them and change them to Short Date. 

Is there a way the data type of the column can be maintained as date on the export? 
I am in Ireland and using standard European based date format (dd/mm/yyyy). 

I searched for a response to this but couldn't find anything and am new to GIT so sorry if I missed it. Thanks!

**Comment by dgosbell** [comment, marked as answer]
> Is there a way the data type of the column can be maintained as date on the export?

So the datatype **is** currently maintained, it's the data format that is not maintained. eg  1 April 2022 and 1/4/2022 (EU) and 4/1/2022 (US) are all the same date, just in different formats. 

The problem with CSV files is that they are just text and they do not have anywhere to store column data types, so the only way to get client tools like Excel to recognize data as a datetime is to use the ISO8601 date format so this is what DAX Studio does.

The problem with using any other format is that you could end up with any of the 4 outcomes below:
1. It may identify the dates correctly
2. It may treat the data as text so none of the sorting/filtering will work correctly
3. It may swap the months and days
4. It may swap the months and days some of the time

And the least common outcome in my experience is option 1.

The best way to export to Excel and maintain the data format would be to use the static Excel output, but that currently has a few bugs when dealing with certain datetime formats, but if I can figure those out then you could export directly to Excel and maintain both the datatype and formatting.

**Reply by JHCairn** [reply]
Thanks @dgosbell. 

How or where would I 'use the ISO8601 date format'? Is that in the PowerBI Desktop or something I do within DAX Studio? 
I have seen other threads where desired 'date' data is being output/exported as 'datetime'. I understand that might bloat the file size, but in my scenario that might not be too bad. Even getting to that outcome I think would be an improvement. If you can suggest how I would  implement that, I would appreciate it. 

As a reference, my source files are mostly JSON files that are updated weekly. I do a bunch of merging, cleaning, and processing the files in PowerBI to generate a weekly report and interactive visuals for analysis. I also export transaction level detailed data from the model using DAX Studio (to static XLSX). These are the files that that have roughly 5-8 columns containing date data that export as general, that have to be changed to date format by the users each week. As noted, the date columns in PowerBI and in DAX Studio do show as the dates.  

Thanks!

**Reply by JHCairn** [reply]
Actually after a bit of a google, I think I have an approach forward. Thank you for the pointer. I will consider this answered and mark it as such.

**Reply by dgosbell** [reply]
> How or where would I 'use the ISO8601 date format'?

You don't have to do anything. When you export to CSV this is the format that is used for any datetime based columns and is basically a format like "yyyy-mm-dd HH:MM:SS". You could see this if you opened the files in something like notepad.

**Comment by JHCairn** [comment]
Actually my dates look like this in Excel:
![image](https://user-images.githubusercontent.com/33914698/164444672-d2d35d90-22a7-4d6d-b7dc-4a013ed89c88.png)

**Reply by JHCairn** [reply]
trying again with a zoomed in screenshot 
![image](https://user-images.githubusercontent.com/33914698/164445006-8d62512b-0df9-46ac-ad11-957bc4d580cb.png)

**Reply by JHCairn** [reply]
I created a new column where I formatted the date value (dd/mm/yyyy) so it reads in the ISO8601 format of YYYY-MM-DD, but now it is text, and when i try to make it a date, it changes is back to dd/mm/yyyy, and on the export to XLS from Dax Studio. 

Below might help. Left column is what is exported from DaxStudio, and right column is what I get when I manually change the column in Excel from General to short date. My users have to select the 5-10 columns with dates each week and manually change them from General to short date in Excel. My goal is to give them output that they don't have to do this with the date columns manually each week. 

![image](https://user-images.githubusercontent.com/33914698/164446403-149cdb65-2694-49f8-a6e7-577702dc5b1d.png)

**Comment by JHCairn** [comment]
It looks like this is simply Excel trying to be helpful. When I did output to file from dax studio instead of output to statis xls, the date columns did show as understandable dates in the CSV files. Unfortunately it gives me 20 different files instead of one excel file with 20 sheets. 

With the excel file and multiple sheets, I am able to rename the sheets and split them apart to separate XLSX files using a short VBA script.

I guess I can look for a different approach to rename the individual files. Can I somehow tell DAX Studio what to name the file outputs from within the script? That could work possibly.

**Comment by dgosbell** [comment]
> It looks like this is simply Excel trying to be helpful.

No this is actually a bug in the Static Excel output. You just had me confused initially as the discussion title and your original post both referred to the CSV output which is different.

There are some differences between the custom formatting codes for DAX and Excel. eg DAX uses 'MM' for the 2 digit month while excel uses 'mm' so I need to figure out some mappings to make this work properly

**Comment by JHCairn** [comment]
Thanks @dgosbell. Sorry I didn't express my issue accurately enough initially, but glad we got there eventually. Do you think this is something that would get fixed at some point? I will keep an eye out for a fix.

**Reply by dgosbell** [reply]
That's OK, the screenshots helped me to understand the issue and it should not be too hard to fix, the tricky bit is just making sure I have all the possibilities covered. I should be able to get a fix for this out shortly.

I also have some ideas around being able to control the naming of files when running multiple queries, but that is probably a longer term thing.
